pkg/operator/ceph/controllers/clusterdisruption/add.go

```diff
@@ -0,0 +1,170 @@
+limitations under the License.
+*/
+
+package clusterdisruption
+
+import (
+	"github.com/rook/rook/pkg/operator/ceph/controllers/controllerconfig"
+	"github.com/rook/rook/pkg/operator/ceph/controllers/nodedrain"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// Add adds a new Controller based on clusterdisruption.ReconcileClusterDisruption and registers the relevant watches and handlers
+func Add(mgr manager.Manager, opts *controllerconfig.Options) error {
+
+	// Add the cephv1 scheme to the manager scheme
+	mgrScheme := mgr.GetScheme()
+	cephv1.AddToScheme(mgr.GetScheme())
+
+	// this will be used to associate namespaces and cephclusters.
+	sharedClusterMap := &ClusterMap{}
+
+	reconcileClusterDisruption := &ReconcileClusterDisruption{
+		client:              mgr.GetClient(),
+		scheme:              mgrScheme,
+		options:             opts,
+		clusterMap:          sharedClusterMap,
+		osdCrushLocationMap: &OSDCrushLocationMap{Context: opts.Context},
+	}
+	reconciler := reconcile.Reconciler(reconcileClusterDisruption)
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return err
+	}
+
+	// enqueues with an empty name that is populated by the reconciler.
+	// There is a one-per-namespace limit on CephClusters
+	enqueueByNamespace := &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(obj handler.MapObject) []reconcile.Request {
+			// The name will be populated in the reconcile
+			namespace := obj.Meta.GetNamespace()
+			if len(namespace) == 0 {
+				logger.Errorf("enqueByNamespace recieved an obj without a namespace: %+v", obj)
+				return []reconcile.Request{}
+			}
+			req := reconcile.Request{NamespacedName: types.NamespacedName{Namespace: namespace}}
+			return []reconcile.Request{req}
+		}),
+	}
+
+	// Watch for CephClusters
+	err = c.Watch(&source.Kind{Type: &cephv1.CephCluster{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for PodDisruptionBudgets and enqueue the CephCluster in the namespace
+	err = c.Watch(
+		&source.Kind{Type: &policyv1beta1.PodDisruptionBudget{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(func(obj handler.MapObject) []reconcile.Request {
+				_, ok := obj.Object.(*policyv1beta1.PodDisruptionBudget)
+				if !ok {
+					// not a pod, returning empty
+					logger.Errorf("PDB handler recieved non-PDB")
+					return []reconcile.Request{}
+				}
+				labels := obj.Meta.GetLabels()
```

Comment:
Is there any reason to prefer predicates? Looks like you have to implement one function each for Create/Update/Delete/Generic or use them all as wrappers for one. With EnqueueRequestsFromMapFunc{}, you can use one function to handle all.

Reply:
Oh, I just realized I can do it with just an Update method and predicate.Funcs

Reply:
Turns out I can just use predicate.Funcs and only implement the Update function, but on further examination, the reason I used a handler is because I'm leaving the Name blank in the request and only populating the Namespace. I've explained why in my other comment about enqueueByNamespace.

The name is populated in the Reconcile() function from a sharedClusterMap object that remembers the name of the CephCluster in the namespace as rook-ceph assumes every ceph object in the namespace belongs to that CephCluster. When the CephCluster is enqueued with handler.EnqueueRequestForObject{}, the sharedClusterMap is updated.